agent/service_manager.go

```diff
@@ -0,0 +1,237 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/structs"
+	"golang.org/x/net/context"
+)
+
+// The ServiceManager is a layer for service registration in between the agent
+// and the local state. Any services must be registered with the ServiceManager,
+// which then maintains a long-running watch of any globally-set service or proxy
+// configuration that applies to the service in order to register the final, merged
+// service configuration locally in the agent state.
+type ServiceManager struct {
+	services map[string]*serviceConfigWatch
+	agent    *Agent
+
+	sync.Mutex
+}
+
+func NewServiceManager(agent *Agent) *ServiceManager {
+	return &ServiceManager{
+		services: make(map[string]*serviceConfigWatch),
+		agent:    agent,
+	}
+}
+
+// AddService starts a new serviceConfigWatch if the service has not been registered, and
+// updates the existing registration if it has. For a new service, a call will also be made
+// to fetch the merged global defaults that apply to the service in order to compose the
+// initial registration.
+func (s *ServiceManager) AddService(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, source configSource) error {
+	reg := serviceRegistration{
+		service:  service,
+		chkTypes: chkTypes,
+		persist:  persist,
+		token:    token,
+		source:   source,
+	}
+
+	// If a service watch already exists, update the registration. Otherwise,
+	// start a new config watcher.
+	s.Lock()
+	watch, ok := s.services[service.ID]
+	s.Unlock()
+	if ok {
+		if err := watch.updateRegistration(&reg); err != nil {
+			return err
+		}
+		s.agent.logger.Printf("[DEBUG] agent.manager: updated local registration for service %q", service.ID)
+	} else {
+		// This is a new entry, so get the existing global config and do the initial
+		// registration with the merged config.
+		args := structs.ServiceConfigRequest{
+			Name:         service.Service,
+			Datacenter:   s.agent.config.Datacenter,
+			QueryOptions: structs.QueryOptions{Token: s.agent.config.ACLAgentToken},
+		}
+		if token != "" {
+			args.QueryOptions.Token = token
+		}
+		var resp structs.ServiceConfigResponse
+		if err := s.agent.RPC("ConfigEntry.ResolveServiceConfig", &args, &resp); err != nil {
+			s.agent.logger.Printf("[WARN] agent: could not retrieve central configuration for service %q: %v",
```

Comment:
suggest we use the terminology we expose in the api and make this `could not retrieve service_defaults config for service %q`